python/ray/util/client/dataclient.py

```diff
@@ -184,17 +185,17 @@ def _process_response(self, response: Any) -> None:
         if response.req_id in self.asyncio_waiting_data:
             can_remove = True
             try:
-                # NOTE: calling self.asyncio_waiting_data.pop() results
-                # in the destructor of ClientObjectRef running, which
-                # calls ReleaseObject(). So self.asyncio_waiting_data
-                # is accessed without holding self.lock. Holding the
-                # lock shouldn't be necessary either.
-                callback = self.asyncio_waiting_data.get(response.req_id)
+                callback = self.asyncio_waiting_data[response.req_id]
                 if isinstance(callback, ChunkCollector):
                     can_remove = callback(response)
                 elif callback:
                     callback(response)
                 if can_remove:
+                    # NOTE: calling del self.asyncio_waiting_data results
+                    # in the destructor of ClientObjectRef running, which
+                    # calls ReleaseObject(). So self.asyncio_waiting_data
+                    # is accessed without holding self.lock. Holding the
```

Comment:
Just to clarify, should ReleaseObject not be called while holding a lock?

Reply:
Yes, I think it's to avoid a problem with object refs being cleaned up while holding locks causing deadlock.